src/jwk.rs

```diff
@@ -0,0 +1,151 @@
+        let thumbprint = Self::calculate_thumbprint(&public_key)?;
+
+        Ok(JsonWebKey {
+            intended_use: Self::intended_use_for_key_pair_type(kp_type),
+            key_id: thumbprint,
+            public_key,
+            private_key: Some(BASE64URL_NOPAD.encode(sk.as_bytes())),
+            ..Default::default()
+        })
+    }
+
+    pub fn from_public_key(source: &str) -> Result<Self> {
+        let (prefix, bytes) = from_public_key(source)?;
+        let vk = VerifyingKey::from_bytes(&bytes)?;
+        let public_key = BASE64URL_NOPAD.encode(vk.as_bytes());
+        let thumbprint = Self::calculate_thumbprint(&public_key)?;
+
+        Ok(JsonWebKey {
+            intended_use: Self::intended_use_for_key_pair_type(&KeyPairType::from(prefix)),
+            key_id: thumbprint,
+            public_key,
+            ..Default::default()
+        })
+    }
+
+    fn intended_use_for_key_pair_type(typ: &KeyPairType) -> String {
+        match typ {
+            KeyPairType::Server
+            | KeyPairType::Cluster
+            | KeyPairType::Operator
+            | KeyPairType::Account
+            | KeyPairType::User
+            | KeyPairType::Module
+            | KeyPairType::Service => "sig".to_owned(),
+            KeyPairType::Curve => "enc".to_owned(),
+        }
+    }
+
+    /// For details on how fingerprints are calculated, see: https://datatracker.ietf.org/doc/html/rfc7638#section-3.1
+    /// For OKP specific details, see https://datatracker.ietf.org/doc/html/draft-ietf-jose-cfrg-curves-06#appendix-A.3
+    pub fn calculate_thumbprint(public_key: &str) -> Result<String> {
+        // We use BTreeMap here, because the order needs to be lexicographically sorted:
+        // https://datatracker.ietf.org/doc/html/rfc7638#section-3.3
+        let components = BTreeMap::from([
+            ("crv", JWK_SUBTYPE),
+            ("kty", JWK_KEY_TYPE),
+            ("x", public_key),
+        ]);
+        let value = json!(components);
+        let mut bytes: Vec<u8> = Vec::new();
+        serde_json::to_writer(&mut bytes, &value).map_err(|_| {
+            err!(
+                ThumbprintCalculationFailure,
+                "unable to serialize public key"
+            )
+        })?;
+        let mut hasher = Sha256::new();
+        hasher.update(&*bytes);
+        let hash = hasher.finalize();
+        Ok(BASE64URL_NOPAD.encode(&hash))
+    }
+}
+
+impl Default for JsonWebKey {
+    fn default() -> Self {
+        Self {
+            intended_use: Default::default(),
+            key_type: JWK_KEY_TYPE.to_string(),
+            key_id: Default::default(),
+            algorithm: JWK_ALGORITHM.to_string(),
+            subtype: JWK_SUBTYPE.to_string(),
+            public_key: Default::default(),
+            private_key: None,
+        }
+    }
+}
+
+impl TryFrom<KeyPair> for JsonWebKey {
+    type Error = crate::error::Error;
+
+    fn try_from(value: KeyPair) -> std::prelude::v1::Result<Self, Self::Error> {
```

Comment:
Nit: I'd import `Result` up at the top of the file as `StdResult`, but up to you

Reply:
Good either way here since the impl is fairly well contained and only used here

Reply:
I meant to use crate::Result here since we define one and this can produce one of the crate::error::Errors, but somehow friendly neighborhood auto-completer picked `std::prelude::v1::Result` instead 😂